core/src/main/scala/org/apache/spark/api/python/PythonWorkerFactory.scala

```diff
@@ -68,6 +100,135 @@ private[spark] class PythonWorkerFactory(pythonExec: String, envVars: Map[String
+
+      while (ze != null) {
+        breakable {
+
+          if (ze.isDirectory()) {
+            // continue
+            break;
+          }
+
+          val fileName = ze.getName();
+          val newFile = new File(outputFolder + File.separator + fileName);
+
+          logDebug("file unzip : " + newFile.getAbsoluteFile());
+
+          // create folders
+          new File(newFile.getParent()).mkdirs();
+
+          val fos = new FileOutputStream(newFile);
+
+          var len: Int = zis.read(buffer);
+
+          while (len > 0) {
+
+            fos.write(buffer, 0, len)
+            len = zis.read(buffer)
+          }
+
+          fos.close()
+        }
+        ze = zis.getNextEntry()
+      }
+
+      zis.closeEntry()
+      zis.close()
+
+    } catch {
+      case e: IOException => logError("exception caught: " + e.getMessage)
+    }
+
+  }
+
+  /**
+   * Create virtualenv using native virtualenv or conda
+   *
+   * Native Virtualenv:
+   *   -  Execute command: virtualenv -p pythonExec --no-site-packages virtualenvName
+   *   -  if wheelhouse specified:
+   *        - Execute command: python -m pip --cache-dir cache-dir install -r requirement_file.txt
+   *      else:
+   *        - Execute command: python -m pip --cache-dir cache-dir install --use-wheel \
+   *                                  --no-index --find-links=wheelhouse -r requirement_file.txt
+   *
+   * Conda
+   *   -  Execute command: conda create --name virtualenvName --file requirement_file.txt -y
+   *
+   */
+  def setupVirtualEnv(): Unit = {
+    logDebug("Start to setup virtualenv...")
+    virtualEnvName = "virtualenv_" + conf.getAppId + "_" + WORKER_Id.getAndIncrement()
+    // use the absolute path when it is local mode otherwise just use filename as it would be
+    // fetched from FileServer
+    val pyspark_requirements =
+      if (Utils.isLocalMaster(conf)) {
+        virtualRequirements
+      } else {
+        virtualRequirements.split("/").last
+      }
+
+    val createEnvCommand =
+      if (virtualEnvType == "native") {
+        if (virtualEnvSystemSitePackages) {
+          Arrays.asList(virtualEnvPath, "-p", pythonExec, "--system-site-packages", virtualEnvName)
+        }
+        else {
+          Arrays.asList(virtualEnvPath, "-p", pythonExec, virtualEnvName)
+        }
+      } else {
+        // Conda
+        Arrays.asList(virtualEnvPath,
+          "create", "--prefix", System.getProperty("user.dir") + "/" + virtualEnvName,
+          "--file", pyspark_requirements, "-y")
+      }
+    execCommand(createEnvCommand)
+    // virtualenv will be created in the working directory of Executor.
+    virtualPythonExec = virtualEnvName + "/bin/python"
+    if (virtualEnvType == "native") {
+      var basePipArgs = mutable.ListBuffer[String]()
+      basePipArgs += (virtualPythonExec, "-m", "pip", "install", "-r", pyspark_requirements)
+      if (!virtualWheelhouse.isEmpty) {
+        unzipWheelhouse("wheelhouse.zip", "wheelhouse")
```

Comment:
wheelhouse.zip is hard coded, it's better to specify it through configuration rather than hard code it and add it through --files.